timely/src/progress/broadcast.rs

```diff
@@ -108,16 +125,33 @@ impl<T:Timestamp+Send> Progcaster<T> {
 
             let addr = &mut self.addr;
             let channel = self.channel_identifier;
-            self.logging.as_ref().map(|l| l.log(crate::logging::ProgressEvent {
-                is_send: false,
-                source: source,
-                seq_no: counter,
-                channel,
-                addr: addr.clone(),
-                // TODO: fill with additional data
-                messages: Vec::new(),
-                internal: Vec::new(),
-            }));
+            self.logging.as_ref().map(|l| {
+
+                let mut messages = Vec::with_capacity(changes.len());
+                let mut internal = Vec::with_capacity(changes.len());
+
+                // TODO: Reconsider `String` type or perhaps re-use allocation.
+                for ((location, time), diff) in recv_changes.iter() {
```

Comment:
I've done this locally as well to track down a potential race, and ended up just changing `messages` and `internal` to a single vector of `changes`, as this filtering can be easily done at the receiver. Any reason you prefer to keep them separate?

Reply:
Oh, only that it was written that way to begin I think, and I wanted to change minimal things. Did you make the change because this was error-prone? Historically they were separate because that's how the data were presented (separate `messages` and `internal`).

Reply:
I think also historically vectors of enums where kept as multiple demux'd vectors to minimize branching in the middle of hot loops (back when the time to update an operator was sub-microsecond. I think that time may have drifted up enough that this sort of attention to detail is less critical.